src/plugins/index_pattern_field_editor/public/components/field_editor/field_editor.tsx

```diff
@@ -5,11 +5,245 @@
  * compliance with, at your election, the Elastic License or the Server Side
  * Public License, v 1.
  */
-import React from 'react';
+import React, { useEffect } from 'react';
+import { i18n } from '@kbn/i18n';
+import { EuiFlexGroup, EuiFlexItem, EuiSpacer, EuiComboBoxOptionOption } from '@elastic/eui';
 
-// eslint-disable-next-line @typescript-eslint/no-empty-interface
-export interface Props {}
+import {
+  Form,
+  useForm,
+  FormHook,
+  UseField,
+  TextField,
+  useFormData,
+  RuntimeType,
+} from '../../shared_imports';
+import { Field } from '../../types';
 
-export const FieldEditor = (props: Props) => {
-  return <div>Placeholder for the future field editor</div>;
+import { RUNTIME_FIELD_OPTIONS } from './constants';
+import { schema } from './form_schema';
+import { getNameFieldConfig } from './lib';
+import { ShadowingFieldWarning } from './shadowing_field_warning';
+import {
+  TypeField,
+  CustomLabelField,
+  ScriptField,
+  FormatField,
+  PopularityField,
+} from './form_fields';
+import { FormRow } from './form_row';
+import { AdvancedParametersSection } from './advanced_parameters_section';
+
+export interface FieldEditorFormState {
+  isValid: boolean | undefined;
+  isSubmitted: boolean;
+  submit: FormHook<Field>['submit'];
+}
+
+export interface FieldFormInternal extends Omit<Field, 'type'> {
+  type: Array<EuiComboBoxOptionOption<RuntimeType>>;
+  __meta__: {
+    isCustomLabelVisible: boolean;
+    isValueVisible: boolean;
+    isFormatVisible: boolean;
+    isPopularityVisible: boolean;
+  };
+}
+
+export interface Props {
+  /** Link URLs to our doc site */
+  links: {
+    runtimePainless: string;
+  };
+  /** Optional field to edit */
+  field?: Field;
+  /** Handler to receive state changes updates */
+  onChange?: (state: FieldEditorFormState) => void;
+  /** Optional context object */
+  ctx?: {
+    /**
+     * An array of field names not allowed.
+     * e.g we probably don't want a user to give a name of an existing
+     * runtime field (for that the user should edit the existing runtime field).
+     */
+    namesNotAllowed?: string[];
```

Comment:
I think we might need to modify this as there can be thousands of fields in an index pattern. I think it would be better to check `indexPattern.getFieldByName` so we don't need to create such a large array. 

on a similar note, I'm not sure we need `existingConcreteFields` as users will be able to define runtime fields on existing fields. I think they should edit that field instead.

Reply:
> I think we might need to modify this as there can be thousands of fields in an index pattern. I think it would be better to check indexPattern.getFieldByName so we don't need to create such a large array.

Not sure I see the difference between providing a handler to get the array when the component mounts vs `namesNotAllowed: indexPattern.getFieldByName()`. In both cases the array will be in memory when opening the flyout.

> I'm not sure we need existingConcreteFields as users will be able to define runtime fields on existing fields. I think they should edit that field instead

Not sure I follow. This array is for the autocomplete of painless syntax (list of fields to chose from the `doc` object) + showing a shadowing warning if a user creates a runtime field with the same name as an existing concrete field. We can discuss it later if you want.